src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
             }
             else
             {
-                emitResult = compilation.Emit(outputAssembly);
+                using (var fs = new FileStream(outputAssembly, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None))
+                {
+                    emitResult = compilation.Emit(peStream: fs, options: _emitOptions);
+                }
+
                 if (emitResult.Success)
                 {
-                    if (PassThru)
+                    if (!CompileOnly)
                     {
                         Assembly assembly = Assembly.LoadFrom(outputAssembly);
-                        CheckTypesForDuplicates(assembly);
-                        WriteTypes(assembly);
+
+                        CacheNewTypes(newTypes);
+                        CacheAssemply(assembly);
+
+                        if (PassThru)
+                        {
+                            WriteTypes(assembly);
+                        }
                     }
                 }
             }
 
-            if (emitResult.Diagnostics.Length > 0)
+            HandleCompilerErrors(emitResult.Diagnostics);
+        }
+
+        private void HandleCompilerErrors(ImmutableArray<Diagnostic> compilerDiagnostics)
+        {
+            if (compilerDiagnostics.Length > 0)
             {
-                HandleCompilerErrors(GetErrors(emitResult.Diagnostics));
+                bool IsError = false;
+
+                foreach (var diagnisticRecord in compilerDiagnostics)
+                {
+                    // We shouldn't specify input and output files in ExtendedOptions parameter
+                    // so suppress errors from Roslyn default command line parcer:
+                    //      CS1562: Outputs without source must have the /out option specified
+                    //      CS2008: No inputs specified
+                    //      BC2008: No inputs specified
+                    if (diagnisticRecord.IsSuppressed ||
```

Comment:
Maybe the check of `diagnisticRecord.IsSuppressed` should be put at the outmost of this method. If `isSuppressed` is true, no need to go through the loop.

Reply:
Sorry, I don't understand - we get diagnisticRecord in the loop - how we can move it out of the loop?